File: src/SalaryMapper.java
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Mapper;

public class SalaryMapper extends Mapper<LongWritable, Text, Text, Text> {
    private final Text jobTitle = new Text();
    private final Text salaryInfo = new Text();
    private final Pattern salaryPattern = Pattern.compile("\\$?(\\d+)K?\\s*-\\s*\\$?(\\d+)K?");
    
    @Override
    public void map(LongWritable key, Text value, Context context) throws IOException, InterruptedException {
        try {
            String line = value.toString();
            String[] fields = line.split(",(?=([^\"]*\"[^\"]*\")*[^\"]*$)");  // CSV splitting accounting for quoted fields
            
            // Skip malformed lines or header
            if (fields.length < 15) {
                return;
            }
            
            // Extract job title (column 15, index 14)
            String title = fields[14].trim();
            if (title.isEmpty() || title.equalsIgnoreCase("job title")) {
                return;  // Skip empty titles or header row
            }
            
            // Extract salary range (column 4, index 3)
            String salaryField = fields[3].trim();
            double avgSalary = parseSalary(salaryField);
            
            if (avgSalary > 0) {
                jobTitle.set(title);
                // Emit salary as: avgSalary,avgSalary,avgSalary,1 (for min,max,avg,count)
                salaryInfo.set(avgSalary + "," + avgSalary + "," + avgSalary + ",1");
                context.write(jobTitle, salaryInfo);
            }
        } catch (Exception e) {
            // Skip any problematic lines
        }
    }
    
    private double parseSalary(String salaryStr) {
        try {
            Matcher matcher = salaryPattern.matcher(salaryStr);
            if (matcher.find()) {
                double min = Double.parseDouble(matcher.group(1));
                double max = Double.parseDouble(matcher.group(2));
                
                // Convert to actual salary if K is in the string
                if (salaryStr.contains("K")) {
                    min *= 1000;
                    max *= 1000;
                }
                
                return (min + max) / 2;
            }
        } catch (Exception e) {
            // Return 0 for unparseable salaries
        }
        return 0;
    }
}

File: python_version/salary_mapper.py
#!/usr/bin/env python3
import sys
import csv
import re

# Regex to capture salary ranges like "$50K - $80K" or "50000 - 80000"
salary_pattern = re.compile(r'\$?(\d+)K?\s*-\s*\$?(\d+)K?')

def parse_salary(salary_str):
    try:
        match = salary_pattern.search(salary_str)
        if match:
            min_salary = float(match.group(1))
            max_salary = float(match.group(2))
            if 'K' in salary_str.upper():
                min_salary *= 1000
                max_salary *= 1000
            return (min_salary + max_salary) / 2
    except:
        pass
    return 0

for line in sys.stdin:
    try:
        # Use csv.reader to handle quoted fields and commas inside quotes
        reader = csv.reader([line])
        fields = next(reader)

        if len(fields) < 15:
            continue

        job_title = fields[14].strip()
        if not job_title or job_title.lower() == "job title":
            continue

        salary_field = fields[3].strip()
        avg_salary = parse_salary(salary_field)

        if avg_salary > 0:
            # Output format: key<TAB>value
            # value = min,max,avg,count → all set to avg for mapper
            print(f"{job_title}\t{avg_salary},{avg_salary},{avg_salary},1")

    except Exception:
        continue  # skip malformed lines
